### DISCUSSION TITLE
read_csv + ORDINALITY ?

### DISCUSSION THREAD

**Discussion by jimdickson007** [duckdb/duckdb#13944] [Q&A] [answered]
I would like to assign a record_number to each record in a file ( ~ ORDINALITY ).
Where multiple files are read (globbing), I would like the record_number value to reset per file.
The OS file would remain unmodified.
I suspect I could use `row_number() over ()` - but then need to guarantee that database engine does not sort input.

**Comment by soerenwolfers** [comment, marked as answer]
The functionality you requested exists for parquet files, via the `file_row_number` parameter, see
https://duckdb.org/docs/data/parquet/overview.html

It has previously been requested for CSV at https://github.com/duckdb/duckdb/discussions/9987 .

The row order is preserved by default according to the bottom of the page 
https://duckdb.org/docs/data/csv/overview.html
so your window solution should work,
but it is never quite clear to me how deep the order preservation guarantees go.

**Reply by jimdickson007** [reply]
For additional reference ( different use case but interesting / easily overlooked ) : https://stackoverflow.com/questions/78635682/duckdb-support-for-postgres-function-with-ordinality

**Comment by jimdickson007** [comment]
For those stumbling upon ```row_number() over ()``` technique - it worked brilliantly until it didn't. Not 100% if this is expected behaviour but adding some references that might be useful

References : [Order Preservation](https://duckdb.org/docs/sql/dialect/order_preservation.html) : 

> GROUP BY (in particular, the output order is undefined and the order in which rows are fed into [order-sensitive aggregate functions](https://duckdb.org/docs/sql/functions/aggregates.html#order-by-clause-in-aggregate-functions) is undefined unless explicitly specified in the aggregate function)

Maybe [read_csv: "preservation of insertion order" vs parallel=true: bug? documentation issue? enhancement? · Issue #15098 · duckdb/duckdb](https://github.com/duckdb/duckdb/issues/15098) 

But I get same results even if read_csv() has ```parallel=false```

```SELECT value AS preserve_insertion_order FROM duckdb_settings() WHERE name = 'preserve_insertion_order';```

**Reply by soerenwolfers** [reply]
any reproducible example of what's going wrong?

**Comment by jimdickson007** [comment]
Yes, will attempt to reproduce reliably with anonymous data.
In the meantime, some other info (not sure if any relevant).
I also tried to use ORDER BY {table} as suggested elsewhere (both in partition by and in main select)
Also realised I derive row_number() over () during load (insert ... select from read_csv). Perhaps should load first then derive value.

The value derived is deterministic - on Windows reading file over mapped drive from OneDrive, the 2nd row is always file_row_number = 36.
In another environment (believe AWS EC2 Linux reading file directly from S3 - will confirm), reading [copy of] same file, file_row_number = 98
File is fixed-width, CR LF except last line which is CR only, UTF-8.

Windows and Linux both run DuckDB v1.1.x 

Still owe you a reproducible test case but this may be crucial finding (so far)

```
row_number() over () 			as row_number, -- correct, table row_number  = file record_number
row_number() over (partition by 'a') 	as row_number, -- wrong,   table row_number != file record_number
```
My _partition by_ included a column which happens to have 1 distinct value (mocked with literal above).
Presence of _partition by_ seems to be root cause.

**Reply by jimdickson007** [reply]
Stumbled upon 1.4 feature
https://duckdb.org/docs/stable/sql/query_syntax/from
```
SELECT * 
FROM read_csv('test.csv') WITH ORDINALITY;
```

**Reply by soerenwolfers** [reply]
`WITH ORDINALITY` is being rewritten to a `row_number` window function. As such, the guarantees from https://duckdb.org/docs/stable/sql/dialect/order_preservation apply. If you find that page unclear, comment on https://github.com/duckdb/duckdb-web/issues/5861 what you think is unclear

**Comment by fabrice-etanchaud** [comment]
Hi all !
Does anybody know if WITH ORDINALITY is still giving valid ORDINALITY values when preserve_insertion_order=false ?

**Reply by soerenwolfers** [reply]
`WITH ORDINALITY` is being rewritten to a `row_number` window function. As such, the guarantees from https://duckdb.org/docs/stable/sql/dialect/order_preservation apply. If you find that page unclear, comment on https://github.com/duckdb/duckdb-web/issues/5861 what you think is unclear
